[DOC] Release notes for 2.4 (#3421)
grafana/tempo: docs/sources/tempo/release-notes/v2-4.md

The Tempo team is pleased to announce the release of Tempo 2.4.

This release gives you:

* New features, including multi-tenant queries and experimental TraceQL metrics queries
* Performance enhancements, thanks to the addition of new caching tiers
* Cost savings, thanks to polling improvements that reduce calls to object storage

As part of this release, vParquet3 has also been promoted to the new default storage format for traces. For more about why we’re so excited about vParquet3, refer to [Accelerate TraceQL queries at scale with dedicated attribute columns in Grafana Tempo](/blog/2024/01/22/accelerate-traceql-queries-at-scale-with-dedicated-attribute-columns-in-grafana-tempo/).

Read the [Tempo 2.4 blog post](/blog/2023/11/01/grafana-tempo-2.3-release-faster-trace-queries-traceql-upgrades/) for more examples and details about these improvements.

These release notes highlight the most important features and bugfixes. For a complete list, refer to the [Tempo changelog](https://github.com/grafana/tempo/releases).

## Features and enhancements

The most important features and enhancements in Tempo 2.4 are highlighted below.

### Multi-tenant queries

> Tempo now allows you to query multiple tenants at once. We’ve made multi-tenant queries compatible with streaming ([first released in v2.2]({{< relref "v2-2#get-traceql-results-faster" >}})) so you can get query results as fast as possible. To learn more, refer to [Cross-tenant federation]({{< relref "../operations/cross_tenant_query/" >}}) and [Enable multi-tenancy]({{< relref "../operations/multitenancy/" >}}). [PRs [3262](https://github.com/grafana/tempo/pull/3262), [3087](https://github.com/grafana/tempo/pull/3087)]

Author: ```suggestion
Tempo now allows you to query multiple tenants at once. We’ve made multi-tenant queries compatible with streaming ([first released in v2.2]({{< relref "./v2-2#get-traceql-results-faster" >}})) so you can get query results as fast as possible. To learn more, refer to [Cross-tenant federation]({{< relref "../operations/cross_tenant_query" >}}) and [Enable multi-tenancy]({{< relref "../operations/multitenancy" >}}). [PRs [3262](https://github.com/grafana/tempo/pull/3262), [3087](https://github.com/grafana/tempo/pull/3087)]
```

### TraceQL metrics (experimental)

We’re excited to announce the addition of metrics queries to the TraceQL language. Metric queries extend trace queries by applying a function to trace query results. This powerful feature creates metrics from traces, much in the same way that LogQL metric queries create metrics from logs.

Tempo 2.4 introduces metrics queries to the TraceQL language as an experimental feature.

Metric queries extend trace queries by applying a function to trace query results.

This powerful feature creates metrics from traces, much in the same way that LogQL metric queries create metrics from logs.

> Initially, only `count_over_time` and `rate` are supported.

Reviewer: this seems repeated content from the traceql metrics page? what do you think about just teasing a rate query and then linking to those docs?

also, should we drop mentions of `count_over_time`? i just don't think it makes sense unless you provide a duration (which we don't support yet) b/c you have no idea what the time period the counts are evaluated over. 

cc @mdisibio

Author: I'll see if I can shorten up the section.

Author: So we are saying we only support rate? I am thinking about just dropping the sentence entirely and just teasing an example and linking to the doc page for TraceQL metrics.

Reviewer 2: > I am thinking about just dropping the sentence entirely and just teasing an example and linking to the doc page for TraceQL metrics.
Sounds good! 

> So we are saying we only support rate?
If you run a query with "count_over_time" it will work and return an answer, so in that regard its "supported". The issue is that in the UI its not obvious the range of time the count is executed over. So if you mouse over a point on the line and you see the value `53`, you know the count of spans is `53`, but you don't know "was that 53 spans seen over a minute? over 30 seconds? over 10 minutes?". With rate, you know that the unit is "spans/second" so if you saw "53" you'd know it was "53 spans per second".

Author: thank you for the explanation. when in doubt, leave it out...

![Metrics visualization in Grafana](/media/docs/tempo/metrics-explore-sample-2.4.png)

For more information, refer to the [TraceQL metrics]({{< relref "../operations/traceql-metrics" >}}) documentation. [PRs [3227](https://github.com/grafana/tempo/pull/3227) [#3252](https://github.com/grafana/tempo/pull/3252), [3258](https://github.com/grafana/tempo/pull/3258)]

To learn more about the TraceQL syntax, see the [TraceQL documentation]({{< relref "../traceql" >}}). For information on planned future extensions to the TraceQL language, refer to [future work]({{< relref "../traceql/architecture" >}}).

#### TraceQL performance improvements

In addition, we’ve improved performance for queries and structural operators

* Improve TraceQL regex performance in certain queries. [PR  [3139](https://github.com/grafana/tempo/pull/3139)]
* Improve TraceQL performance in complex queries. [[PR 3113](https://github.com/grafana/tempo/pull/3113)]
* TraceQL/Structural operators performance improvement. [[PR 3088](https://github.com/grafana/tempo/pull/3088)]

### vParquet3 is now the default block format

> Tempo 2.4 makes [vParquet3]({{< relref "../configuration/parquet" >}}) default, a Parquet version designed to be more compatible with other Parquet implementations, available as a production-ready option. This block format improves query performance relative to previous formats. [[PR 2526](https://github.com/grafana/tempo/pull/2526)]

Reviewer: > a Parquet version designed to be more compatible with other Parquet implementations

i don't think this is true. we should just double down on the addition of dedicated columns which was the real reason for vp3. should we link to the blogpost as well as the docs?

Author: We have a link to the blog post earlier in the release notes, but you're right that it should be here too. 

```suggestion
Tempo 2.4 makes [vParquet3]({{< relref "../configuration/parquet" >}}) default.
Dedicated attribute columns, available using vParquet3, improve query performance by storing the largest and most frequently used attributes in their own columns, rather than in the generic attribute key-value list.
For more information, refer to the 
[Accelerate TraceQL queries at scale with dedicated attribute columns in Grafana Tempo](/blog/2024/01/22/accelerate-traceql-queries-at-scale-with-dedicated-attribute-columns-in-grafana-tempo/) blog post. [[PR 2526](https://github.com/grafana/tempo/pull/2526)]
```

Author: not using my suggestino in favor of Jen's.

> Tempo 2.4 makes [vParquet3]({{< relref "../configuration/parquet" >}}) default, a Parquet version designed to be more compatible with other Parquet implementations, available as a production-ready option. This block format improves query performance relative to previous formats. [[PR 2526](https://github.com/grafana/tempo/pull/2526)]

Author: I've incorporated this change into my change above.

We’re excited about [vParquet3]({{< relref "../configuration/parquet" >}}) relative to prior formats because of its support for [dedicated attribute columns]({{< relref "../operations/dedicated_columns" >}}), which help speed up queries on your largest and most queried attributes.

> 

Author: ```suggestion

Dedicated attribute columns, available using vParquet3, improve query performance by storing the largest and most frequently used attributes in their own columns, rather than in the generic attribute key-value list.
For more information, refer to 
[Accelerate TraceQL queries at scale with dedicated attribute columns in Grafana Tempo](/blog/2024/01/22/accelerate-traceql-queries-at-scale-with-dedicated-attribute-columns-in-grafana-tempo/). [[PR 2526](https://github.com/grafana/tempo/pull/2526)]

```

* Added per-tenant configurable `remote_write` headers to metrics-generator. [#3175](https://github.com/grafana/tempo/pull/3175)
* Added variable expansion support to overrides configuration. [PR 3175](https://github.com/grafana/tempo/pull/3175)
* Added HTML pages `/status/overrides` and `/status/overrides/{tenant}`. [PR 3244](https://github.com/grafana/tempo/pull/3244) [#3332](https://github.com/grafana/tempo/pull/3332)
* Precalculate and reuse the vParquet3 schema before opening blocks. [PR 3367](https://github.com/grafana/tempo/pull/3367)
* Made the trace ID label name configurable for remote written exemplars. [PR 3074](https://github.com/grafana/tempo/pull/3074)
* Performance improvements in span filtering by introducing `AttributePolicyMatch` & `IntrinsicPolicyMatch` structures to match span attributes based on strongly typed values & precompiled regexp. [PR 3025](https://github.com/grafana/tempo/pull/3025)
* Merged the processors overrides set through runtime overrides and user-configurable overrides. [PR 3125](https://github.com/grafana/tempo/pull/3125)

> * Removed experimental websockets support for search streaming. GRPC is the supported method of streaming results. [PR 3307](https://github.com/grafana/tempo/pull/3307)

Author: Moved to the upgrade section and added a section for it.

* Introduced localblocks process configuration option to select only server spans. [PR 3303](https://github.com/grafana/tempo/pull/3303)

> * Localblocks processor honor tenant max trace size limit. [PR 3305](https://github.com/grafana/tempo/pull/3305)

Author: Removed.

## Upgrade considerations

> When [upgrading](../setup/upgrade/) to Tempo 2.4, be aware of these considerations and breaking changes.

Author: ```suggestion
When [upgrading]({{< relref "../setup/upgrade" >}}) to Tempo 2.4, be aware of these considerations and breaking changes.
```

### Transition to vParquet 3

vParquet3 format is now the default block format. It is production ready and we highly recommend switching to it for improved query performance and [dedicated attribute columns]({{< relref "../operations/dedicated_columns" >}}).

Upgrading to Tempo 2.4 modifies the Parquet block format. Although you can use Tempo 2.3 with vParquet2 or vParquet3, you can only use Tempo 2.4 with vParquet3.

> With this release, vParquet is being deprecated. Tempo 2.4 will still read vParquet1 blocks. However, Tempo will fail to configure them. [[PR 3377](https://github.com/grafana/tempo/pull/3377/files#top)]

Author: We never called it vParquet1, though. it was vParquet and then vParquet2. I've added some context above. What do you think?

> With this release, vParquet is being deprecated. Tempo 2.4 will still read vParquet1 blocks. However, Tempo will fail to configure them. [[PR 3377](https://github.com/grafana/tempo/pull/3377/files#top)]

Author: I've added the rewritten section above.

For information on upgrading, refer to [Upgrade to Tempo 2.4](https://grafana.com/docs/tempo/latest/setup/upgrade/) and [Choose a different block format](https://grafana.com/docs/tempo/next/configuration/parque#choose-a-different-block-format) .

> ### Updated, removed, or renamed configuration parameters

Reviewer: These fields have all been removed in favor of the new cache config:

```
storage:
  trace:
    cache:
    search:
      cache_control:
    background_cache:
    memcached:
    redis:
```

maybe mention them here with a "see below for details"? up to you

Author: What do you think about adding the a new row to the table and including these with a link to the section below?

   </td>
  </tr>
  <tr>
   <td><code>tempo_query_frontend_queries_total{op="searchtags|metrics"}</code>
   </td>
   <td>Removed deprecated frontend metrics configuration option
   </td>
  </tr>
</table>

The distributor now returns 200 for any batch containing only `trace_too_large` and `max_live_traces` errors. The number of discarded spans are still reflected in the `tempo_discarded_spans_total metrics`.

### Cache configuration refactored

The major cache refactor to allow multiple role-based caches to be configured. [[PR 3166](https://github.com/grafana/tempo/pull/3166)]
This change resulted in the following fields being deprecated.
These have all been migrated to a top level `cache:` field.

For more information about the configuration, refer to the [Cache]({{< relref "../configuration#cache" >}}) section.

The old configuration block looked like this:

```yaml
storage:
  trace:
    cache:
    search:
      cache_control:
    background_cache:
    memcached:
    redis:
```

With the new configuration, you create your list of caches,- with either `redis` or `memcached` cluster with your config, then define the types of data and roles.

```yaml

> cache:

Reviewer: some of these settings are super esoteric. i would use something simpler like i put in the blogpost:

```
cache:
  caches:
  - memcached:
      host: <some memcached cluster>
    roles:
    - bloom
    - parquet-footer
  - memcached:
      host: <some memcached cluster>
    roles:
    - frontend-search
```

Author: Okay. I'll swap out the example with the one you provided.

Author: Done. Kept the short example here and the longer example in the upgrade page.

### TraceQL metrics (experimental)

We’re excited to announce the addition of metrics queries to the TraceQL language. Metric queries extend trace queries by applying a function to trace query results. This powerful feature creates metrics from traces, much in the same way that LogQL metric queries create metrics from logs.

Tempo 2.4 introduces metrics queries to the TraceQL language as an experimental feature.

Metric queries extend trace queries by applying a function to trace query results.

This powerful feature creates metrics from traces, much in the same way that LogQL metric queries create metrics from logs.

Initially, only `count_over_time` and `rate` are supported.

For example:

```

> 

Author: ```suggestion
```

{ resource.service.name = "foo" && status = error } | rate()

> 

Author: ```suggestion
```

```

In this case, we are calculating the rate of the erroring spans coming from the service `foo`. Rate is a `spans/sec` quantity.

```

> 

Author: ```suggestion
```

> 

Author: ```suggestion
```

{ resource.service.name = "foo" && status = error } | rate() by (span.http.route)

> 

Author: ```suggestion
```

            consistent_hash: true
            host: memcached
            max_idle_conns: 100
            service: memcached-client
            timeout: 200ms
          roles:
            - bloom

        - memcached:
            consistent_hash: true
            host: memcached
            max_idle_conns: 100
            service: memcached-client
            timeout: 200ms
          roles:
            - parquet-footer

        - memcached:
            consistent_hash: true
            host: memcached-frontend-search
            max_idle_conns: 100
            service: memcached-client
            timeout: 50ms
          roles:
            - frontend-search

        - memcached:
            consistent_hash: true
            host: memcached-parquet-page
            max_idle_conns: 100
            service: memcached-client
            timeout: 200ms
          roles:
            - parquet-page
```

## Security fixes

The following vulnerabilities have been addressed:

* Addressed [CVE-2023-5363](https://github.com/advisories/GHSA-xw78-pcr6-wrg8).
* Updated the `memcached` default image in jsonnet for multiple CVEs [PR 3310](https://github.com/grafana/tempo/pull/3310)

> 

Author: ```suggestion
```
